src/DotLiquid/Util/StrFTime.cs

```diff
@@ -16,87 +21,201 @@ public static class StrFTime
+                    dateTimeOffset
+                    ));
+        }
 
-            while (n < pattern.Length)
-            {
-                string s = pattern.Substring(n, 1);
 
-                if (n + 1 >= pattern.Length)
-                    output += s;
-                else
-                    output += s == "%"
-                        ? Formats.ContainsKey(pattern.Substring(++n, 1)) ? Formats[pattern.Substring(n, 1)].Invoke(dateTime) : "%" + pattern.Substring(n, 1)
-                        : s;
-                n++;
-            }
+        /// <summary>
+        /// Formats a date-time for a single specifier within the requested format. For ease of processing
+        /// the specifier is broken down into it's consituent parts of: flags, width, directive.
+        ///
+        /// A specifier consists of a percent (%) character, zero or more flags, optional minimum field width and a conversion directive as follows:
+        /// <param name="orig">the entire specifier, such as %Y for a 4-digit year</param>
+        /// <param name="flags">zero or more flags to change formatting, such as '0' (zero padding).</param>
+        /// <param name="width">optional minimum field width</param>
+        /// <param name="directive">The required data value, such as Y for a 4-digit year</param>
+        /// <param name="obj">the source date-time object</param>
+        /// </summary>
+        private static String StrFTimeMatchEvaluator(String orig, IEnumerable<String> flags, int? width, String directive, object obj)
+        {
+            var result = orig;
+            directive = PreProcessDirective(directive, flags, width);
+
+            bool isDateTimeOffset = obj is DateTimeOffset;
+            if (OffsetFormats.ContainsKey(directive) && isDateTimeOffset)
+                result = OffsetFormats[directive].Invoke((DateTimeOffset)obj);
+            else if (Formats.ContainsKey(directive))
+                result = Formats[directive].Invoke(isDateTimeOffset ? ((DateTimeOffset)obj).DateTime : (DateTime)obj);
+            else
+                return orig; // This is an unspecified directive
 
-            return output;
+            return flags.ToList().Aggregate(result, (current, flag) => ApplyFlag(flag, (width ?? 2), current));
         }
 
-        /// <summary>
-        /// Formats a date using a ruby date format string
-        /// </summary>
-        /// <param name="dateTime"></param>
-        /// <param name="pattern"></param>
-        /// <returns></returns>
-        public static string ToStrFTime(this DateTimeOffset dateTime, string pattern)
+        // Pre-process the directive for some of the quirkier cases, such as '%:z' and '%3N'.
+        private static String PreProcessDirective(String directive, IEnumerable<String> flags, int? width)
         {
-            var output = new System.Text.StringBuilder();
+            var flagString = String.Join("", flags);
+            if ("z".Equals(directive) && ":".Equals(flagString))

```

Comment:
Equals takes a comparison type. I think you want culture insensitive / ordinal comparison here. But case sensitive, right?

Reply:
Done